man/man1/expound.1
.TH EXPOUND "1" "Feb 2011" "expound" "User Commands"
.SH NAME
expound \- annotates genomic regions
.SH SYNPOSIS
.BI "expound [-o " output_file "] [-O " output_pattern "] 
.BI " [-a " annotation_file " ...] [-A " legacy_annotation_file " ...]"
.BI " [-c " chromosome_file "] [\-covered|\-partial|\-inclusive|\-nearest] "
.BI " [-p " port "] [-H " host "] [-q]"
.BI "" region_file " ..."

.SH DESCRIPTION
.BR expound " (formerly known as " coord2anno )
reads one or more tables that map genomic regions to annotations, then
reads one or more files containing regions, and annotates those regions.
Region files can be
.SM SAM, BAM,
a simple table, or a
.SM FASTA
file that encodes regions in sequence identifiers; coordinates can be
either
.SM UCSC
or
.I Ensembl
style, or a mix of both.

.SH OPTIONS
.IP "\fB-o, \-output \fIoutput_file\fR"
Defines the output file.  If neither \fB\-o\fR nor \fB\-O\fR is given,
output is sent to stdout.

.IP "\fB-O, \-output-pattern \fIpattern\fR"
Defines a pattern for the construction of output file names.  One output
file is created for each input file.  The output file name is obtained
by replacing the strings \fI%p\fR in \fIpattern\fR with the path of the
input file, \fI%n\fR with its name without extension, \fI%e\fR with its
extension, \fI%f\fR with its full name without extension, and \fI%s\fR
with its full name.  The string \fI%%\fR results in a single percent
sign.

.IP "\fB-a, \-annotation \fIannotation_file\fR"
Defines an annotation file.  Annotations must be in BED format and are
read before any other processing starts.  Multiple annotations may be
given, the union of the results is output.

.IP "\fB-A, \-legacy-annotation \fIannotation_file\fR"
Defines an annotation files just like
.BR \-annotation ,
but this file is in the legacy five column format.  Using this option
is discouraged.

.IP "\fB-c, \-chroms \fIchromosome_file\fR"
Defines the chromosome translation table with legacy annotations.  If a
translation table is given,
.B expound
tries to translate
.I Ensembl
names and coordinates in annotations into
.SM UCSC
names and coordinates.

.IP "\fB-s, \-nostrand\fR"
Ignore strand information in region files.  This is primarily useful
when annotating
.SM BAM
files, which always contain strand information, when this information
isn't actually useful.

.IP "\fB\-covered\fR"
Output only annotations that are fully contained in a region.

.IP "\fB\-partial\fR"
Output only annotations that overlap into a region, but are not fully
contained in it.

.IP "\fB\-inclusive\fR"
Output annotations that overlap a region at least partially.  This is
the default.

.IP "\fB\-nearest\fR"
Output partial and total overlaps as with
.BR \-inclusive ,
but also output the nearest annotation for each region that does not
overlap any annotation.  The (signed) distance is appended to the
annotation, separated by an '@' character.

.IP "\fB-h, \-host\fR"
Switch to client mode and and set the server host to connect to.  See
below under
.IR "client-server mode" .

.IP "\fB-p, \-port\fR"
In client mode, set the port to connect to.  Else switch to server mode
and set the port to listen on.  See below under
.IR "client-server mode" .

.IP "\fB-q, \-quiet\fR"
Operate silently, without progress reports.

.IP "\fB-h, -?, \-help, \-usage\fR"
Print a summary of command line options.

.SH "FILES"
.SS "Annotation Table"
The annotation table should be in
.SM BED
format.  For
.B expound
to do anything useful, the
.SM BED
file must have at least 4 columns, since column 4 ("name") is used as
the annotation.  If the
.SM BED
file has 6 or more columns, column 6 ("strand") is used to make the
annotation strand specific.  Empty lines, lines starting with a hash
mark ('#') and track definition lines (those starting with the word
"track") are ignored.

.B expound
also understands an ad-hoc five column format.  This 
.I was a bad idea
and these files should be phased out.  Therefore, they will not be
documented here.  If you are smart enough to make five column tables,
you are smart enough to turn them into
.SM BED
files even without documentation, so please do that.  In the meantime,
they can still be loaded through the
.B \-legacy-annotation
option.


.SS "Chromosome Translation Table"
The chromosome translation table is a text file with one entry per line
and three whitespace separated columns.  The first line can optionally
be a header line starting with '#', which is ignored.  The fields
are:
.IP \[bu] 2
.I Ensembl
.SM ID
(a number or an accession number)
.IP \[bu] 2
.SM UCSC ID
(chr-something)
.IP \[bu] 2
Offset minus one

The offset field is 0 for ordinary chromosomes where just the name needs
to be mapped.  To map contigs with accession numbers to chromosomes
.IR "" "(" "chr_random" " or " "chrUn" "),"
the offset is the zero-based start position in the concatenated
.SM UCSC
style chromosome.  Note that chromosome translation is only applied
to legacy annotation files.

.SS "Region Files"
.B expound
tries to infer the format of region files automatically.  Compressed and
uncompressed
.SM BAM
files are recognized from their header.  If a file starts with a '>', it
is interpreted as
.SM FastA.
If a file starts with an '@', or if the first line has 11 or more fields
and field 5 can be interpreted as a
.SM CIGAR
line, it is interpreted as a
.SM SAM
file.  Anything else is read as a
.SM BED
file.

There used to be support for a custom five-column region file, which has
been dropped in favor of
.SM BED.
If you were smart enough to generate the five-column file, you are smart
enough to convert it to
.SM BED.
You are encouraged to do so.


.SS "FastA File With Regions"
If a region file begins with '>', it is interpreted as a 
.SM FastA
file where the sequence identifiers contain region information.  All
identifiers must contain at least four fields separated by colons (':').
These fields are:
.IP \[bu] 2
chromosome
.IP \[bu] 2
strand
.IP \[bu] 2
start position
.IP \[bu] 2
end position

Abusing 
.SM FastA
in this way
.IR "is a terrible idea" .
If you have such files, you should convert them to 
.SM BAM.

.SS "Output Table"
The output table a text file with one entry per line and two whitespace
separated columns.  The first line starts with '#' and contains a
descriptive header.  The first column contains the region's identifier,
the second column contains all annotations, separated by colons (':').
Regions with no annotation at all are suppressed from the output.


.SH "NOTES"
.SS "Coordinate Systems"
Internally,
.B expound
always represents regions as zero-based, half-open intervals.  The same
convention is used in 
.SM BAM, BED, FastA
and legacy five-column files.  In contrast,
.SM SAM
uses one-based, closed intervals.

Earlier versions supported a brittle heuristic to infer the conventions
used in a particular file.  This heuristic has been dropped; please
convert your legacy files to a standard format and use the conventions
of that format.  The "chromosomes" in legacy annotation files can still
be translated from
.I Ensembl
conventions to
.SM UCSC
conventions by providing a translation table.  In this case, one-based,
closed intervals are translated into zero-based, half-open intervals,
and an optional offset is applied.


.SS "Regions"
In annotation files, the smaller of the two given coordinates is
interpreted as start coordinate, the greater as end coordinate.  No
inference about which strand is annotated is drawn from the relative
magnitude of the coordinates.

In input files, start coordinates must be less than or equal to end
coordinates, otherwise the region is interpreted as empty.

Note that in
.I Ensemble
style coordinates it is impossible to encode an empty region.  In
.SM UCSC
style coordinates, a region with equal start and end coordinates is
interpreted as empty.  No warning or error is given for empty regions,
even though they don't make much sense in the context of
.B expound .

.SS "Strands"
Both annotations and regions have an associated strand, but either may
be unknown.  If an annotation has no known strand, it is understood as
applying to both strands.  If a region has no known strand, annotations
are looked up on both strands and the union of them is reported.

The forward strand is always encoded as "+", the reverse strand is
encoded as "-", and an unknown strand is encoded as "0" or, where
appropriate, a missing field.

.SS "Client-Server Mode"
.B expound
can optionally operate in a client-server mode.  In this mode, the
server effectively acts as an in-memory cache of annotations.  New
annotations can be added at any time by simply running the client,
cached annotations may automatically be evicted from the server at any
time.

The server is started by running 
.B expound
with the
.B \-port
option.  All other options except
.B \-quiet
are ignored and the server will listen for connections on the given
port.

The client is started by running
.B expound
with both the
.B \-host
and the
.B \-port
options.  All other options work exactly as when running in standalone
mode.  If the server already knows about an annotation file, it will be
used as is, else the client will upload it.  Note that the set of
annotations is identified by the file name, so be careful when using
relative pathnames or shell tricks that generate annotations on the fly.


.SH "BUGS"
There is no way to report only annotations that fully contain a region.
Note that annotations fully \fIcontained\fR in a region can be reported.

The output file format is ad-hoc.  A future version of
.B expound
may well drop support for custom tables and swap in a standard format.

Handling of regions with
.I start
greater than
.I end
is inconsistent, but due to historical reasons this is unlikely to get
fixed.

The handling of
.I Ensembl
and
.SM UCSC
coordinate conventions is fragile and difficult to comprehend, but
there's hardly anything we can do about it.


.SH "AUTHOR"
Written by Udo Stenzel <udo_stenzel@eva.mpg.de>.




src/expound.hs
{-# LANGUAGE ExistentialQuantification, Rank2Types #-}
{- This is now supposed to read both Ensembl- and UCSC-style input, and
 - it had a very weird heuristic to convert them into each other.  We
 - drop this shit now, the rules are:  following the specification, all
 - coordinates are 0-based half open intervals for BED, BAM,
 - weird-FastA, 5col; 1-based closed intervals for SAM.  Annotations in
 - 5col format (and only in 5col format!) can be rewritten using a
 - translation table.  If the coordinate conventions change, the offsets
 - in the translation table need to reflect that.
 -
 - Strand information is optional now, too.  We handle it this way:
 -
 - * If the strand field starts with '-', only the reverse strand is meant.
 - * If the strand field is "0" or absent, both strands are meant.
 - * Else only the forward strand is meant.
 - * Every annotation is reported at most once.
 -
 - Internally, we *always* use zero-based, half-open intervals.  This is
 - also the assumption for the 5col format, except when translation
 - from Emsembl to UCSC is requested.  Since Ensembl uses one-based,
 - closed intervals, we'll follow that convention.  If that is
 - confusing, just don't use stupid home grown fole formats. -}

import Bio.Prelude                    hiding ( Word, bracket )
import Bio.Iteratee
import Control.Concurrent.Async
import Control.Monad.Catch                   ( bracket )
import Control.Monad.Trans.RWS.Strict hiding ( listen, reader )
import Data.Binary.Put                       ( runPut )
import Data.Vector                           ( (!) )
import Network.Socket                 hiding ( send )
import Paths_biohazard_tools                 ( version )
import System.Console.GetOpt
import System.IO                             ( IOMode(..), openFile, hClose )

import qualified Data.ByteString.Char8              as S
import qualified Data.HashMap.Strict                as M
import qualified Network.Socket.ByteString.Lazy     as N ( sendAll )

import Diet
import FormattedIO
import Network
import Symtab

-- A function that gives an Iteratee that will consume the annotations
-- for one file and return some sort of summary, and a function that
-- processes the summaries.
data Output = forall u . Output (IO ()) (FilePath -> Iteratee [Annotation] IO u) ([u] -> IO ())

data Options = Options
     { optOutput      :: Output
     , optWhich       :: Which
     , optAnnotations :: forall b . [( Maybe ChromTable -> Enumeratee Bytes [Region] IO b, FilePath )]
     , optChromTable  :: Maybe ChromTable
     , optXForm       :: Region -> Region
     , optProgress    :: String -> IO ()
     , optPort        :: Maybe String
     , optHost        :: Maybe String }

defaultOptions :: Options
defaultOptions = Options
     { optOutput      = Output (print_header stdout) (const $ make_table stdout) (const $ return ())
     , optWhich       = Inclusive
     , optAnnotations = []
     , optChromTable  = Nothing
     , optXForm       = id
     , optProgress    = hPutStr stderr
     , optPort        = Nothing
     , optHost        = Nothing }

options :: [OptDescr (Options -> IO Options)]
options =
     [ Option "o" ["output"]              (ReqArg set_output "FILE") "write output to FILE"
     , Option "O" ["output-pattern"]         (ReqArg set_opat "PAT") "write output to files following PAT"
     , Option [ ] ["summarize"]                (NoArg set_summarize) "only summarize annotations to stdout"
     , Option "a" ["annotation"]           (ReqArg read_anno "FILE") "read annotations from FILE in Bed format"
     , Option "A" ["legacy-annotation"] (ReqArg read_anno_5c "FILE") "read annotations from FILE in legacy 5c format"
     , Option "c" ["chroms"]             (ReqArg read_chroms "FILE") "read chr translation table from FILE"

     , Option "s" ["nostrand"]  (NoArg  set_nostrand) "ignore strand information in region files"
     , Option [ ] ["covered"]   (NoArg   set_covered) "output only annotations that cover the query interval"
     , Option [ ] ["partial"]   (NoArg   set_partial) "output only annotations that cover only part of the query"
     , Option [ ] ["inclusive"] (NoArg set_inclusive) "output annotations that overlap at least part of the query"
     , Option [ ] ["nearest"]   (NoArg   set_nearest) "output the closest annotation if none overlaps" ]
  where
    set_output f opts = return $ opts { optOutput = Output h b t }
      where h   = print_header stdout
            t _ = return ()
            b _ = if f == "-" then make_table stdout
                              else bracket (liftIO $ openFile f WriteMode) (liftIO . hClose)
                                           (\hdl -> liftIO (print_header hdl) >> make_table hdl)

    read_anno_5c f opts = do hPutStr stderr legacy_warning
                             return $ opts { optAnnotations = (read5col, f) : optAnnotations opts }

    read_anno    f opts =    return $ opts { optAnnotations = (const readMyBed,f) : optAnnotations opts }

    read_chroms  f opts = enumFile defaultBufSize f readChromTable >>= run >>=
                          \s -> return $ opts { optChromTable = Just s }

    set_nostrand   opts = return $ opts { optXForm = eraseStrand }
    set_summarize  opts = return $ opts { optOutput = Output (return ()) make_summary print_summary }
    set_opat     f opts = return $ opts { optOutput = Output (return ()) (write_file_pattern f) (const $ return ()) }

    set_covered    opts = return $ opts { optWhich =   Covered }
    set_partial    opts = return $ opts { optWhich =  Fragment }
    set_inclusive  opts = return $ opts { optWhich = Inclusive }
    set_nearest    opts = return $ opts { optWhich =   Nearest }

common_options :: [OptDescr (Options -> IO Options)]
common_options =
    [ Option "p"  ["port"]      (ReqArg set_port "PORT") "listen on PORT or connect to PORT"
    , Option "H"  ["host"]      (ReqArg set_host "HOST") "connect to HOST"
    , Option "q"  ["quiet"]            (NoArg set_quiet) "do not output progress reports"
    , Option "h?" ["help","usage"] (NoArg $ const usage) "display this information"
    , Option "V"  ["version"]               (NoArg vrsn) "display version number and exit" ]
  where
    set_port p opts = return $ opts { optPort     =          Just p }
    set_host h opts = return $ opts { optHost     =          Just h }
    set_quiet  opts = return $ opts { optProgress = \_ -> return () }

    vrsn          _ = do pn <- getProgName
                         hPutStrLn stderr $ pn ++ ", version " ++ showVersion version
                         exitSuccess

usage :: IO a
usage = do pn <- getProgName
           putStrLn (usageInfo (header pn) $ options ++ common_options)
           exitSuccess

legacy_warning :: String
legacy_warning = unlines [ "Warning: The five column format is not a standard format.  Consider"
                         , "         converting your annotations to the standard BED format." ]

header :: String -> String
header pn = unlines [ "Usage: " ++ pn ++ " [OPTION...] files...", ""
                    , "Annotates genomic regions given by coordinates.  Input files"
                    , "and annotation files are BED or BAM files.  Can operate in"
                    , "client-server-mode.  Options are:" ]

type LookupFn = RevSymtab -> Int -> Int -> [IDiet] -> AnnoSet

standardLookup :: (Int -> Int -> IDiet -> [Word]) -> LookupFn
standardLookup which syms s e diets =
    Hits . map ((!) syms . fromIntegral) . unions $ map (which s e) diets

lookupNearest :: LookupFn
lookupNearest syms s e diets =
    let overlaps = unions $ map (lookupI s e) diets
        closeby  = sort $ filter (not . null . snd) $ map (lookupLeft s) diets ++ map (lookupRight e) diets
    in case (overlaps, closeby) of
        ([], [  ]) -> Hits []
        ([], as:_) -> NearMiss (syms ! fromIntegral a) d where (d,a:_) = as
        (is,   _ ) -> Hits $ map ((!) syms . fromIntegral) is


interpretWhich :: Which -> RevSymtab -> Int -> Int -> [IDiet] -> AnnoSet
interpretWhich Covered   = standardLookup lookupICovered
interpretWhich Fragment  = standardLookup lookupIPartial
interpretWhich Inclusive = standardLookup lookupI
interpretWhich Nearest   = lookupNearest


main :: IO ()
main = do
    args <- getArgs
    when (null args) usage

    let (os, files, errors) = getOpt Permute (options ++ common_options) args
    opts <- foldr (>=>) return os defaultOptions
    unless (null errors) $ mapM_ (hPutStr stderr) errors >> exitFailure

    case (optHost opts, optPort opts) of
        (Nothing, Just  p) -> do let (os', files', errors') = getOpt Permute common_options args
                                 opts' <- foldr (>=>) return os' defaultOptions
                                 unless (null errors') $ do mapM_ (hPutStr stderr . (++) "in server mode: ") errors'
                                                            exitFailure
                                 unless (null files') $ do hPutStrLn stderr "no file arguments allowed in server mode"
                                                           exitFailure
                                 run_server p opts'

        (Nothing, Nothing) -> run_standalone opts files
        (Just  h, Just  p) -> run_client h p opts files
        (Just  _, Nothing) -> hPutStrLn stderr "--host given, but no --port" >> exitFailure


run_standalone :: Options -> [FilePath] -> IO ()
run_standalone opts files = do
    emptyD <- unsafeFreezeDiet =<< newDiet
    (!syms,!anno) <- foldr (>=>) run [ enum $= reader (optChromTable opts)
                                     | (reader, filepath) <- optAnnotations opts
                                     , let enum = if filepath == "-"
                                                  then enumFd   defaultBufSize stdInput
                                                  else enumFile defaultBufSize filepath ] $
                        joinI $ progressNum "reading annotations" 16384 (optProgress opts) $
                        ilift execCPS $ (,) <$>
                                readGeneTable <*> (liftIO . mapM unsafeFreezeDiet =<< lift get_anno)
    case optOutput opts of
        Output hdr iter summ -> do
            hdr
            myReadFiles (optProgress opts) files >=> summ $ \file ->
                let annotate (name,c,strs,s,e) =
                            (,) name $ interpretWhich (optWhich opts) syms s e $
                            withSenses strs $ \str -> M.lookupDefault emptyD (str c) anno
                in mapStream (annotate . optXForm opts) =$ iter file


myReadFiles :: (String -> IO ()) -> [FilePath] -> (FilePath -> Iteratee [Region] IO b) -> IO [b]
myReadFiles prg fs it
    | null   fs = (:[]) <$> enum1 "<stdin>" (enumFd defaultBufSize stdInput)
    | otherwise = forM fs $ \f ->
                        if f == "-" then enum1 "<stdin>" (enumFd defaultBufSize stdInput)
                                    else enum1 f (enumFile defaultBufSize f)
  where
    enum1 nm en = en >=> run $ readInput =$ progressNum nm 16384 prg =$ it nm


run_client :: HostName -> ServiceName -> Options -> [FilePath] -> IO ()
run_client h p opts files = do
    let hints = defaultHints { addrFlags = [AI_ADDRCONFIG, AI_CANONNAME] }
    addrs <- getAddrInfo (Just hints) (Just h) (Just p)
    let addr = head addrs
    sock <- socket (addrFamily addr) (addrSocketType addr) (addrProtocol addr)
    connect sock (addrAddress addr)

    -- request/upload each annotation set
    buf' <- reduceM (optAnnotations opts) S.empty $ \buf (reader, fp) ->
                let enum = if fp == "-" then enumFd   defaultBufSize stdInput
                                        else enumFile defaultBufSize fp in
                enum >=> run $
                joinI $ reader (optChromTable opts) $
                joinI $ progressNum ("uploading " ++ fp) 16384 (optProgress opts) $
                ilift (runClient sock buf) $ sendGeneTable fp >> lift get

    -- fork thread to upload all the input
    uploader <- async $ do void $ myReadFiles (optProgress opts) files $ \file -> do
                                lift . N.sendAll sock . runPut . putRequest $ StartFile (S.pack file)
                                mapStreamM_ ( N.sendAll sock . runPut . putRequest
                                            . Anno (optWhich opts) . optXForm opts )
                           N.sendAll sock . runPut $ putRequest EndStream
    link uploader

    -- receive responses, print output
    case optOutput opts of
        Output hdr iter summ -> do
            hdr
            runClient sock buf' >=> summ $ myReadNet iter

    void $ wait uploader
    close sock

  where
    runClient sock incoming cl = fst <$> evalRWST cl sock incoming
    reduceM xs a f = foldM f a xs

    -- Ask the server for an annotation set, upload it if necessary.
    sendGeneTable :: FilePath -> Iteratee [Region] Client ()
    sendGeneTable fp = do
        lift . send . putRequest $ StartAnno (S.pack fp)
        resp <- lift $ receive getResponse
        case resp of
            KnownAnno   -> return ()
            UnknownAnno -> do mapStreamM_ $ \r -> send (putRequest (AddAnno r))
                              lift . send $ putRequest EndAnno
            _           -> fail "the server talks nonsense"

    -- Reads from network and passes it on, calling the continuation
    -- once for each "file".
    myReadNet :: (FilePath -> Iteratee [Annotation] IO b) -> Client [b]
    myReadNet k = do rsp <- receive getResponse
                     case rsp of
                         StartedFile nm -> go [] (k $ unpack nm)
                         EndedStream    -> return []
                         _              -> huh
      where
        huh = fail "unexpected answer from server"

        go acc it = do
            rsp <- receive getResponse
            case rsp of
                Result nm annos -> lift (enumPure1Chunk [(nm,annos)] it) >>= go acc
                StartedFile nm  -> lift (run it) >>= \ !r -> go (r:acc) (k $ unpack nm)
                EndedStream     -> return $ reverse acc
                _               -> huh


type HalfTable = ( S.ByteString, MAnnotab, Symtab )
type FullTable = ( IAnnotab, Symtab, RevSymtab )
type FullTables = M.HashMap S.ByteString FullTable

run_server :: ServiceName -> Options -> IO ()
run_server svname opts = do
    full_tables <- newMVar M.empty
    listener <- socket AF_INET Stream defaultProtocol
    port <- fromIntegral <$> (readIO svname :: IO Int)
    bind listener $ SockAddrInet port iNADDR_ANY
    listen listener 1
    optProgress opts $ "waiting for connections on port " ++ shows port "\27[K\n"
    forever $ do (sk, addr) <- accept listener
                 optProgress opts $ "connection from " ++ shows addr "\27[K\n"
                 let pr m = optProgress opts $ shows addr ": " ++ m
                 void $ forkIO $ void $ service getRequest putResponse (serverfn pr full_tables) sk (Nothing,[])

serverfn :: (String -> IO ())
         -> MVar FullTables
         -> ( Maybe HalfTable, [FullTable] )
         -> Request
         -> IO ( ( Maybe HalfTable, [FullTable] ), Maybe Response)
serverfn _ _fulltables (Just _, _) (StartAnno _) = fail "client tried to open a second annoset"
serverfn pr fulltables (Nothing, fts) (StartAnno name) = do
    tablemap <- readMVar fulltables
    case M.lookup name tablemap of
        Just m  -> do pr $ "brought known annotation " ++ shows name " in scope\n"
                      return ( (Nothing, m:fts), Just KnownAnno)
        Nothing -> do pr $ "starting new annotation set " ++ shows name "\n"
                      return ( (Just (name, M.empty, M.empty), fts), Just UnknownAnno )

serverfn _ _ st (StartFile nm) = return ( st, Just $ StartedFile nm )
serverfn _ _ st  EndStream     = return ( st, Just   EndedStream )

serverfn _ _ (Nothing,_) (AddAnno _) = fail "client tried growing an annotation set without opening it"
serverfn _ _ (Just (name, annotab, symtab), st) (AddAnno (gi, chrom, strs, s, e)) =
    runCPS (findSymbol gi >>= \val ->
                    sequence_ $ withSenses strs $ \str -> findDiet (str chrom) >>=
                            liftIO . addI (min s e) (max s e) (fromIntegral val))
           (\_ symtab' annotab' -> return ((Just (name, annotab', symtab'), st), Nothing))
           symtab annotab

serverfn _ _ (Nothing,_) EndAnno = fail "client tried ending an annotation set without opening it"
serverfn pr fulltables (Just (name, annotab, symtab), s) EndAnno = do
    annotab' <- sequenceA $ M.map unsafeFreezeDiet annotab
    let anno' = (annotab', symtab, invertTab symtab)
    tablemap <- takeMVar fulltables
    putMVar fulltables $! M.insert name anno' tablemap
    pr $ "memoized new annotation set " ++ shows name "\n"
    return ((Nothing, anno':s), Nothing)

serverfn _ _ st@(_,annosets) (Anno which (name,c,strs,s,e)) = do
    emptyD <- unsafeFreezeDiet =<< newDiet
    let result = Result name . foldr combine_anno_sets (Hits []) $
           [ interpretWhich which syms s e (withSenses strs $ \str -> M.lookupDefault emptyD (str c) anno)
           | (anno, _, syms) <- annosets ]
    return (st, Just result)

combine_anno_sets :: AnnoSet -> AnnoSet -> AnnoSet
combine_anno_sets (Hits []) x = x
combine_anno_sets x (Hits []) = x
combine_anno_sets (Hits xs) (NearMiss _ _) = Hits xs
combine_anno_sets (NearMiss _ _) (Hits ys) = Hits ys
combine_anno_sets (Hits xs) (Hits ys)      = Hits (xs++ys)
combine_anno_sets (NearMiss a d) (NearMiss b d')
    | abs d <= abs d' = NearMiss a d
    | otherwise       = NearMiss b d'




